File: packages/amap_kit_map/amap_kit_map_android/android/src/main/kotlin/com/github/amapkit/map/overlays/heatmap/HeatmapsController.kt
package com.github.amapkit.map.overlays.heatmap

import com.amap.api.maps.AMap
import com.github.amapkit.map.Convert
import com.github.amapkit.map.PlatformHeatmap
import com.github.amapkit.map.PlatformHeatmapUpdates

/** Owns every heatmap tile overlay belonging to one map instance. */
class HeatmapsController(private val map: AMap) {
    private val heatmapControllers = mutableMapOf<String, HeatmapController>()

    fun update(updates: PlatformHeatmapUpdates) {
        removeHeatmaps(updates.toRemove)
        changeHeatmaps(updates.toChange)
        addHeatmaps(updates.toAdd)
    }

    fun dispose() {
        heatmapControllers.values.forEach(HeatmapController::remove)
        heatmapControllers.clear()
    }

    fun addHeatmaps(heatmaps: List<PlatformHeatmap>) {
        heatmaps.forEach(::addHeatmap)
    }

    private fun changeHeatmaps(heatmaps: List<PlatformHeatmap>) {
        heatmaps.forEach { heatmap ->
            heatmapControllers[heatmap.id]?.let { controller ->
                Convert.interpretHeatmapOptions(heatmap, controller)
                controller.applyChanges()
            } ?: addHeatmap(heatmap)
        }
    }

    private fun removeHeatmaps(ids: List<String>) {
        ids.forEach { id -> heatmapControllers.remove(id)?.remove() }
    }

    private fun addHeatmap(heatmap: PlatformHeatmap) {
        val builder = HeatmapBuilder()
        val id = Convert.interpretHeatmapOptions(heatmap, builder)
        heatmapControllers.remove(id)?.remove()
        heatmapControllers[id] = HeatmapController(map, builder.build())
    }
}

File: packages/amap_kit_map/amap_kit_map_ios/ios/amap_kit_map_ios/Sources/amap_kit_map_ios/HeatmapsController.swift
import CoreLocation
import MAMapKit
import UIKit

/// Maintains every heatmap belonging to one map instance, mirroring the
/// Android per-object controller pattern.
final class HeatmapsController {
  private unowned let mapView: MAMapView
  private var controllers: [String: HeatmapController] = [:]

  init(mapView: MAMapView) {
    self.mapView = mapView
  }

  func update(_ updates: PlatformHeatmapUpdates) throws {
    for value in updates.toAdd {
      try add(value)
    }
    for value in updates.toChange {
      try change(value)
    }
    updates.toRemove.forEach(remove)
  }

  func renderer(for overlay: MAHeatMapTileOverlay) -> MATileOverlayRenderer? {
    controllers.values.first(where: { $0.overlay === overlay })?.renderer(for: overlay)
  }

  func dispose() {
    controllers.values.forEach { $0.remove() }
    controllers.removeAll()
  }

  private func add(_ value: PlatformHeatmap) throws {
    remove(value.id)
    let controller = try HeatmapController(mapView: mapView, value: value)
    controllers[value.id] = controller
    if value.visible { controller.attach() }
  }

  private func change(_ value: PlatformHeatmap) throws {
    guard let controller = controllers[value.id] else {
      try add(value)
      return
    }
    let wasVisible = controller.value.visible
    try controller.update(value)
    if wasVisible != controller.value.visible {
      if controller.value.visible {
        controller.attach()
      } else {
        controller.remove()
      }
    }
  }

  private func remove(_ id: String) {
    controllers.removeValue(forKey: id)?.remove()
  }
}
